Updated glint derivatives and glint solution for the analytical line

evan-greenbrg opened the pull request:
Updates in this PRT:

1. Further abstracted the surface model out of the radiative transfer class. This involves moving the derivatives of radiance with respect to the surface state elements into the surface class (e.g. for the [multicomponent surface](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/surface/surface_multicomp.py#L263-L290), for the [glint_model_surface](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/surface/surface_glint_model.py#L166-L198)). The [dRT functions](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/radiative_transfer/radiative_transfer.py#L445-L486) solely deal with the RT statevector elements. 

2. RT functions now take explicit arguments. Calls of `get_L_coupled` and `get_L_down_transmitted` are performed in a [helper function](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/radiative_transfer/radiative_transfer.py#L407-L443), which should handle switch off between the 1c and 4c model. Throughout `forward.py`, we now similarly call the RT and surface quantities before passing them into the respective derivative and `calc_rdn` function calls. See [calc_meas](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/core/forward.py#L253-L293), [K](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/core/forward.py#L327-L403), and [Kb](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/core/forward.py#L405-L465) functions.

3. The analytical line algorithm is simplified. We allow the user to initialize the AOE routine with 3 possible values  [(invert_simple, invert_algebraic, and the super pixel value)](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/utils/analytical_line.py#L362-L370).  We also use the same subroutine across all surface models. This relies on the [similar construction](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/inversion/inverse_simple.py#L307-L317) of the matrix, `H`, that is a property of the surface class. For example, [the glint model](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/surface/surface_glint_model.py#L200-L251), which inherits the linearization of the `rfl` terms from the `multicomponent_surface`.

4. This updates the linearization of the 4c model used in the AOE solution. The [multicomponent surface](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/surface/surface_multicomp.py#L311), and the [glint surface](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/surface/surface_glint_model.py#L240).

6. [Edited glint derivatives to reflect the new 4c model](https://github.com/evan-greenbrg/isofit/blob/c3ee11c272b832472d6a07e22f1dd884ded0db17/isofit/surface/surface_glint_model.py#L138-L164). This results in a much improved optimization compared to dev that reflects both the prior and the measurement.

Other edits:
- The outputs for the analytical line will be separated by surface reflectance and surface non_reflectance components. All possible output files are for example: 
`prm20221026t182807_rfl`
`prm20221026t182807_uncert`
`prm20221026t182807_surf_non_rfl`
`prm20221026t182807_surf_non_rfl_uncert`

evan-greenbrg committed:
Updates to clean inverse_analytical and updated glint derivatives

Some cleaning for consistency

Use fresnel function in surface model

Moved drfl_dsurface into the surface model

cleaning

Move fresnel function out of radiative_transfer

set background rfl and initialize analytical line with background rfl

Analytical Line uses superpixel as background rfl and init values

Move Seps and Sa into the loop

Update to reflect new linear factorization

Glint analytical line now uses atm_interp for sky glint

Merge with dev

Abstract analytical model to the surface modules

reverting removing lower bounds on rfl inversion

added missing idx_surface attributes

Inverse_simple now backwards compatible

STARTED astracting surface model calls in RT with updated derivatives

Abstracted the interp check in analytical line

Formatting

evan-greenbrg commented:
A non-water example with the multicomponent_surface:

<img width="1550" alt="Screenshot 2025-02-05 at 1 18 57 PM" src="https://github.com/user-attachments/assets/222c42e0-0e12-4973-b0dc-bd2895efcd9a" />

unbohn commented on isofit/inversion/inverse_simple.py:
```diff
@@ -231,7 +231,8 @@ def invert_analytical(
         geom: geometry object corresponding to given measurement
-        x_RT: the radiative transfer state variables
+        x0: the initialization state including surface from the superpixel
+            and the the atm from the smoothed atmosphere.
```
@evan-greenbrg Here's one 'the' too much.

evan-greenbrg replied:
Fixed!

unbohn commented on isofit/inversion/inverse_simple.py:
```diff
@@ -253,151 +254,117 @@ def invert_analytical(
-        geom,
-    )
-    # x_alg contains [rfl_est, Ls_est, coeffs]
+    x = np.copy(x0)
```
@evan-greenbrg Why do you switch from `x0` to `x` here? Is there any reason for not keeping the `x0` nomenclature?

evan-greenbrg replied:
I changed the variable name to make it clear that the x0 function argument is the initial condition, and that x evolves during the iteration. x0 doesn't need to be kept unchanged, but I like the variable name switch. Can change the copy statement to a simple `x = x0`.

unbohn replied:
Understood! Let's follow your suggestion.

unbohn commented on isofit/inversion/inverse_simple.py:
```diff
@@ -253,151 +254,117 @@ def invert_analytical(
+    # Atmospheric spherical albedo and total trans up for radiance model
+    r = fm.RT.get_shared_rtm_quantities(x_RT, geom)
+    s = r["sphalb"]
+    t_total_up = r["transm_up_dir"] + r["transm_up_dif"]
```
@evan-greenbrg `t_total_up` is not used in subsequent lines, so we could perhaps remove this one.

evan-greenbrg replied:
Removed!

unbohn commented on isofit/inversion/inverse_simple.py:
```diff
@@ -253,151 +254,117 @@ def invert_analytical(
-                full_xk[outside_ret_windows] = outside_ret_const
-            x[fm.idx_surface] = full_xk
-            trajectory.append(x)
+    # Get the doward radiance. Should be stationary with fixed atm and geom
```
@evan-greenbrg Typo in the docstring.

evan-greenbrg replied:
Fixed!

unbohn commented on isofit/radiative_transfer/radiative_transfer.py:
```diff
@@ -440,51 +445,55 @@ def drdn_dRT(
 
         # atmospheric spherical albedo
         s_alb = r["sphalb"]
+        t_total_up = r["transm_up_dir"] + r["transm_up_dif"]
```
@evan-greenbrg You moved this up, so perhaps adding its own docstring again instead of putting it directly under the spherical albedo. Just for clarity.

evan-greenbrg replied:
Should be resolved with new edits.

unbohn commented on isofit/surface/surface_glint_model.py:
```diff
@@ -132,3 +149,80 @@ def summarize(self, x_surface, geom):
+        return H
+
+    @staticmethod
+    def fresnel_rf(vza):
```
@evan-greenbrg Very good move! It makes total sense to have this function as a surface class attribute.

unbohn commented on isofit/surface/surface_glint_model.py:
```diff
@@ -132,3 +149,80 @@ def summarize(self, x_surface, geom):
+        """
+        Linearization of the glint terms to use in AOE inner loop.
+        Function will fetch the linearization of the rho terms and
+        add the matrix coponents for the direct glint term.
```
@evan-greenbrg One more typo.

evan-greenbrg replied:
Fixed!

unbohn commented on isofit/surface/surface_glint_model.py:
```diff
@@ -132,3 +149,80 @@ def summarize(self, x_surface, geom):
+        rho_ls = self.fresnel_rf(geom.observer_zenith)
+        # Direct component holding dif component constant
+        g_dir = rho_ls * (L_down_dir / (L_down_dir + L_down_dif))
+        g_dif = rho_ls * (L_down_dif / (L_down_dir + L_down_dif))
```
@evan-greenbrg We call this calculation of `g_dir` and `g_dif` quite a few times at different locations in the code. Maybe moving it into a separate function?

evan-greenbrg replied:
Great idea, I added a new function that can be called throughout glint_model_surface.

```python
    def glint_spectrum(self, geom, L_down_dir, L_down_dif):
        """Calculates the sun (dir) and sky (dif) glint spectrums"""
        rho_ls = self.fresnel_rf(geom.observer_zenith)
        # direct sky transmittance
        g_dir = rho_ls * (L_down_dir / (L_down_dir + L_down_dif))
        # diffuse sky transmittance
        g_dif = rho_ls * (L_down_dif / (L_down_dir + L_down_dif))

        return g_dir, g_dif
```

unbohn replied:
Nice! In the docstring, do you mean 'spectra' or 'spectrum'?

evan-greenbrg replied:
Changed to spectra.

unbohn commented on isofit/surface/surface_lut.py:
```diff
@@ -179,3 +179,36 @@ def summarize(self, x_surface, geom):
+        Linearization of the surface reflectance terms to use in the
+        AOE inner loop (see Susiluoto, 2025). We set the quadratic
+        spherical albedo term to a constant background, which makes
+        simplifies the linearization
```
@evan-greenbrg One more typo.

evan-greenbrg replied:
Fixed!

unbohn commented on isofit/surface/surface_multicomp.py:
```diff
@@ -235,3 +237,28 @@ def summarize(self, x_surface, geom):
+        Linearization of the surface reflectance terms to use in the
+        AOE inner loop (see Susiluoto, 2025). We set the quadratic
+        spherical albedo term to a constant background, which makes
+        simplifies the linearization
```
@evan-greenbrg See previous comment.

unbohn commented on isofit/utils/analytical_line.py:
```diff
@@ -261,12 +271,17 @@ def __init__(
+        self.atm_bands = []
+
+        # How many iterations to use for invert_analytical
+        self.num_iter = 1
```
@evan-greenbrg Is it safe to put `num_iter = 1` here?

evan-greenbrg replied:
Good question. I added this to allow for the flexibility if we wanted to run the inverse_analytical algo with more iterations than 1. We don't have this ability on the dev-branch. Looking at it again, perhaps it's a better idea to add it as an optional Worker argument?

unbohn replied:
Yeah, I think that would actually be better.

evan-greenbrg committed:
Reformatted fm.K and broke out derivatives in surface models

Remove depreciation question

evan-greenbrg committed:
Fixed bugs that came up during examples check

Debug on analytical line

evan-greenbrg committed:
PR edits

Change name of glint_spectra

add in copy to inverse_analytical

evan-greenbrg commented:
I messed up the order on rebasing #637 641 in relation to #637. The commit order here reflects the commit order:
 `dev` -> #641 -> #637 while based on conversations the commit order should be:  `dev` -> #637 -> #641.

I've spent some time trying to work through options to get the `dev` -> #637 -> #641 within this PR, and while it is possible, would unnecessarily introduce a more complicated commit history and opportunities for mistakes in the merge. I'm keeping this PR as is for now with the note that I have a backup branch: https://github.com/evan-greenbrg/isofit/tree/backup/clean/analytical_line with the correct commit order, which makes it easy to roll back parts of #641 as needed.

unbohn commented:
@evan-greenbrg Thanks for the update! This kind of problem sounds familiar to me. Every rebasing is a new experience ;)

evan-greenbrg committed:
Explicitely pass into calc_rdn, wrapper fncts to aggregate surface and RT quantities in fm

Passing tests

typos

evan-greenbrg committed:
Fixed derivative updating RT quantities

Updated drfl_dglint

Updated drfl_dsurface and perturbation

Annotations

Reverting derivative changes

Added note on fixed apparent rfl in dRT, and testing slow example

evan-greenbrg committed:
Fixed incorrect check for 1c vs 4c model within RT.calc_rdn

evan-greenbrg committed:
Split out upsampling for vectors and matrices

evan-greenbrg commented:
Results from comparing Dev-branch to _this_-branch solutions for a sample EMIT scene (Cuprite: emit20240626t165035). These solutions reflect the 4c radiance model with a multicomponent_surface in both cases.

![emit20240626t165035_compare](https://github.com/user-attachments/assets/34a5460e-cad6-4d1d-9e04-ff5f8cde7579)

evan-greenbrg committed:
Updating for cluster

evan-greenbrg committed:
Fixed typo in linear radiance model

unbohn committed:
Added nan-interpolation to isofit.py to avoid no data pixels in reflectance output.

evan-greenbrg committed:
Added ability to use different AOE initialization strategies

evan-greenbrg committed:
Added error handling for NaNs during AOE iteration

evan-greenbrg committed:
Merge branch 'dev' of github.com:evan-greenbrg/isofit into clean/analytical_line

evan-greenbrg committed:
Updated arguments to kwargs in ForwardModel

evan-greenbrg committed:
Analytical line outputs are split between rfl and non_rfl states
